### ISSUE TITLE
Positive click on mask / negative click on background

### ISSUE THREAD

**Issue by dfrumkin** [alibaba/ClickSEG#2]
During benchmarking every click is either a positive on the background or a negative on the foreground.  This results in the assumption that for a click at (y, x), pred_mask[y, x] is different from previous_mask[y, x] at https://github.com/alibaba/ClickSEG/blob/main/isegm/inference/evaluation.py#L50
Since progressive merge is not performed during training, it is run only on a fully trained model where the above assumption always (?) holds - for the Clicker that you defined that is.

Two bugs / features with the above:

1. During a real user interaction, nothing prevents the user from placing a positive click on an existing mask or a negative click on the background.  Then progressive merge is going to ruin the mask.  For example, for a freshly loaded image without a previous mask, if the user places a negative click anywhere in the image, the resulting mask will cover the whole image.

2. Another possibility is that the user's correction simply did not lead to any changes - at least at the location of the click.  I have actually run into such a situation while playing with your model and real user clicks, but in theory there's no reason why this can't happen even with ideal center-of-the-region clicks you use for your benchmarks.

The fix for (1) would be not to run the algorithm at all.  Since you don't have this scenario in your test and don't provide inference code, I guess it's irrelevant.  The fix for (2) would probably be to detect this strange result and discard the predicted mask logging an error/warning.

I ran into such scenarios when testing your code in inference mode.  I guess for benchmarking it's a very minor issue; however, for inference / real user interaction, the results may be really surprising.

**Comment by XavierCHEN34** [comment]
Sorry for late reply. I only maintain the repo forked into my own account.

1. The first situation would not ruin the mask at all. When the user place a negative click as the first, the new prediction would be "all zero", but not "mask will cover the whole image". Similarly, when user puts a click on the region that already been segmented as foreground, the mask would keep unchanged.

2. In the practical application, **when user put a click on the foreground mask, we would calculate a b-box according to the size of the whole mask and positions of all click, then we only update the mask within this local region**.
However, we made user study. The result show that: the annotators think "placing a positive click on the foreground mask also changes the mask" makes them confusing; They hope to strictly follow the rule of "add clicks on wrong region". Hence, we remove this strategy.

In general, FocalClick is a flexible tool, you could make minor modifications to fit your needs.

**Comment by dfrumkin** [comment]
Thank you for your response, Xi!

To be more specific, I meant that the code at https://github.com/alibaba/ClickSEG/blob/main/isegm/inference/evaluation.py#L50 makes an assumption that `previous_mask` and `corr_mask` are different at y, x.  This would normally happen when the user wants to change the mask value at y, x and the change indeed happens.  Practically, though, there is no guarantee, and if `previous_mask[y, x] == corr_mask[y, x]`, the result of this code will be different from what one would expect.

This has happened to me in two different scenarios that I described above and I changed my code to fix it.
